operators/src/adapters/raster_subquery_adapter.rs

```diff
@@ -316,6 +320,266 @@ where
     }
 }
 
+#[pin_project(project=StateInnerProjection)]
+#[derive(Debug, Clone)]
+enum StateInner<A, B, C> {
+    CreateNextQuery,
+    RunningQuery {
+        #[pin]
+        query: A,
+        query_rect: QueryRectangle<SpatialPartition2D>,
+    },
+    RunningFold(#[pin] B),
+    ReturnResult(Option<C>),
+    Ended,
+    Error,
+}
+
+type StateInnerForClippy<'a, P, FoldFuture, FoldMethod, TileAccu> = StateInner<
+    QueryFuture<'a, P>,
+    RasterFold<'a, P, FoldFuture, FoldMethod, TileAccu>,
+    RasterTile2D<P>,
+>;
+
+/// This adapter allows to generate a tile stream using sub-querys.
+/// This is done using a `TileSubQuery`.
+/// The sub-query is resolved for each produced tile.
+#[pin_project(project = RasterSubQueryAdapterProjection2)]
+pub struct RasterSubQueryAdapter2<'a, PixelType, RasterProcessorType, SubQuery>
+where
+    PixelType: Pixel,
+    RasterProcessorType: RasterQueryProcessor<RasterType = PixelType>,
+    SubQuery: SubQueryTileAggregator<PixelType>,
+{
+    /// The `RasterQueryProcessor` to answer the sub-queries
+    source_processor: &'a RasterProcessorType,
+    /// The `QueryContext` to use for sub-queries
+    query_ctx: &'a dyn QueryContext,
+    /// The `QueryRectangle` the adapter is queried with
+    query_rect_to_answer: RasterQueryRectangle,
+    /// The `GridBoundingBox2D` that defines the tile grid space of the query.
+    grid_bounds: GridBoundingBox2D,
+
+    /// The `SubQuery` defines what this adapter does.
+    sub_query: SubQuery,
+
+    /// This `TimeInterval` is the time currently worked on
+    current_time_start: TimeInstance,
+    current_time_end: Option<TimeInstance>,
+    /// The `GridIdx2D` currently worked on
+    current_tile_spec: TileInformation,
+
+    /// This current state of the adapter
+    #[pin]
+    state: StateInnerForClippy<
+        'a,
+        PixelType,
+        SubQuery::FoldFuture,
+        SubQuery::FoldMethod,
+        SubQuery::TileAccu,
+    >,
+}
+
+impl<'a, PixelType, RasterProcessor, SubQuery>
+    RasterSubQueryAdapter2<'a, PixelType, RasterProcessor, SubQuery>
+where
+    PixelType: Pixel,
+    RasterProcessor: RasterQueryProcessor<RasterType = PixelType>,
+    SubQuery: SubQueryTileAggregator<PixelType>,
+{
+    #[allow(clippy::too_many_lines)]
+    fn next_step(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Result<Option<RasterTile2D<PixelType>>>>> {
+        let mut this = self.project();
+
+        // check if we ended in a previous call
+        if matches!(*this.state, StateInner::Ended) {
```

Comment:
i do not think that a loop is needed here and it will add complexity that is not needed.
For each tile we either run through all states to produce a tile OR we return early with and error.

The state is not really a state machine that is correct. What it does is what async .wait does internally. 
Since we use futures here we might return Poll:Pending and in that case we need to store our state. async .await does something similar. it generates a state enum where the scope of the current .await is represented. Sadly for streams that is not possible.

Since there are no state transitions other then the chain a -> b -> c -> d , a -> d and * -> ended i think the current code is more readable. The ifs are only needed since we might skip b and c.